particles/Tutorial/Javascript/Demo/source/RandomComputer.js

```diff
@@ -0,0 +1,42 @@
+/**
+ * @license
+ * Copyright 2019 Google LLC.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+/* global defineParticle */
+
+defineParticle(({DomParticle, html}) => {
+
+  const template = html`
+<b>Player One</b><input value="{{name}}" placeholder="Enter your name" spellcheck="false" on-change="onNameInputChange">
+  `;
+
+  return class extends DomParticle {
+
+    get template() {
+      return template;
+    }
+
+    update({gameState}, {}) {
+      if (gameState) {
+        
+        const emptyCells = [];
+        const board = gameState.board.split(`,`);
+        // Determine which cells are empty.
+        for (let i = 0; i < board.length; i++) {
+          if (board[i] == ``) {
+            emptyCells.push((i + 1).toString());
+          }
+        }
+        // Select a random empty cell.
+        const selection = Math.floor(Math.random() * emptyCells.length);
+        this.updateSingleton('computerMove', {move: emptyCells[selection]});
```

Comment:
It looks to me like computer will play a move any time the board changes. But it is not always its turn. So it is up to Game to decide which moves to accept. This sounds a bit like a race condition to me and an opportunity for weird behaviour.

I would prefer a design where game can tell the player "Hey, it's your move" and then the player, be it a human or a computer, can make a move.

Right now I'm not sure if you would put a delay of a few seconds in a computer, wouldn't the game allow the human to make a few moves in a row? Maybe it would, maybe it would not, I'm not sure. But the fact that I'm not sure is not good. I don't like that you had to put a lot of special casing in the Game to make computer stop polluting the board, it is a symptom of this design that appears suboptimal to me.

We could have a feedback channel from the Game to players through a "currentPlayer" handle, or have one handle per player with a "Is it my turn?" bit, or the Game could clear the move it receives and when it is time for a given player to make a move it can put in a move with a `?` or `-1` instead of the move as a single for that player that it is time for them to put in a correct move. These suggestions are roughly in my order of preference, but if you have even better one, that's great.

What do you think?

Reply:
Also when it is not a Human's move, the board should be locked. We could display this as some background color, or simply not having `cursor:pointer` and `:hover` CSS, which will make the board appear non interactive I think.

Reply:
We could put a 200ms-500ms or so delay in the computer to give the impression of computer thinking and showcase that the board is locked at that time.

Reply:
I haven't done the UI changes yet, but the backend changes are in.